### DISCUSSION TITLE
General approach to multi-column pagination?

### DISCUSSION THREAD

**Discussion by jedwards1211** [prisma/prisma#19040] [Q&A]
Let's say I have
```
model User {
  id              String @id @default(autoincrement())
  name            String
}
```

And I want to paginate over `name`.  It's not unique, so I'd need to include `id` to make the pagination deterministic.  With raw SQL, cursor-based pagination is as simple as:

```sql
CREATE INDEX User_name_id_idx ON "User" USING btree (name, id);
-- $1, $2 come from cursor passed to API
SELECT * FROM "User" WHERE (name, id) > ($1, $2) ORDER BY name, id LIMIT 100;
```

Does Prisma support this?  Disappointingly the docs say the cursor must be a "unique, sequential column", so it doesn't sound like it.  But this is a very common use case so it seems like a surprising limitation.

If there's an index on `(name, id)` then the Prisma could hypothetically support:

```ts
const secondQueryResults = await prisma.user.findMany({
  skip: 1,
  take: 100,
  cursor: {
    name,
    id,
  },
  orderBy: {
    name: 'asc',
    id: 'asc',
  },
})
```
Generating SQL `WHERE (name, id) > ($1, $2)` (for Postgres at least) as above.

Another less performant approach would be to still use `id` as the cursor and generate SQL like
```
SELECT * FROM "User" WHERE (name, id) > (
  SELECT name, id FROM "User" WHERE id = $1
) ORDER BY name, id LIMIT 100;
```

Then at least we can fit within the current "unique, sequential column" limitation.  But it's not clear as someone evaluating Prisma if it can do either of these things.

**Comment by nurul3101** [comment]
Hello @jedwards1211 👋 

Would defining an id by combining `id` and `name` like this
```prisma
model User {
  id   Int    @default(autoincrement())
  name String

  @@id([id, name])
}
```

And querying the users like this work for you?
```ts
  const secondQueryResults = await prisma.user.findMany({
    skip: 1,
    take: 100,
    cursor: {
      id_name: {
        id: 1,
        name: 'name',
      },
    },
    orderBy: {
      name: 'asc',
      id: 'asc',
    },
  });
```

**Reply by jedwards1211** [reply]
Any way that works is better than nothing, though I would be much happier with being able to pass `cursor: { id: 1, name: 'name' }`.  I think Prisma could have a better design here.  This has various problems:

- Why not use `@@index([id, name])` instead, since you'd typically create an index for this, not a unique constraint
- `cursor` could have the same properties as `orderBy` as would be expected, technical limitations aren't a good reason it should be different
- I wouldn't typically refer to the combination of these two columns as an "id", because the `name` serves no purpose in uniquely identifying a row.
- The docs say composite `@@id` doesn't support Mongodb, but multicolumn indexes and pagination are possible with Mongodb

**Reply by nurul3101** [reply]
Thank you for sharing the feedback, I'll discuss this internally with the team 👍

**Reply by jedwards1211** [reply]
Sounds good.  I want to emphasize that this is probably the most common use case for pagination.  For example take a basic twitter clone: you have to paginate tweets by `(date, id)`.  And you definitely don't think of creating a unique constraint on `(date, id)`, you just think of that as an index.

**Reply by jedwards1211** [reply]
@nurul3101 also I was just trying out `@@id` and I can only use it once, meaning I can only paginate one multi-column index.  Databases support more than one multi-column index and prisma pagination should too.
